pkg/tsdb/cloudmonitoring/resource_handler.go

```diff
@@ -233,10 +243,74 @@ func processData(res *http.Response, responseFn processResponse) ([]byte, int, e
 		}
 	}
 	if err != nil {
-		return nil, http.StatusInternalServerError, fmt.Errorf("unable to encode response %v", err)
+		return nil, fmt.Errorf("unable to encode response %v", err)
+	}
+	return buf.Bytes(), nil
+}
+
+func processData(data io.ReadCloser, encoding string, responseFn processResponse) ([]json.RawMessage, string, int, error) {
+	body, err := decode(encoding, data)
+	if err != nil {
+		return nil, "", http.StatusBadRequest, fmt.Errorf("unable to decode response %v", err)
+	}
+
+	response, token, err := responseFn(body)
+	if err != nil {
+		return nil, "", http.StatusInternalServerError, fmt.Errorf("data processing error %v", err)
+	}
+	return response, token, 0, nil
+}
+
+func doRequest(req *http.Request, cli *http.Client, responseFn processResponse) (string, http.Header, []json.RawMessage, string, int, error) {
+	res, err := cli.Do(req)
+	if err != nil {
+		return "", nil, nil, "", http.StatusBadRequest, err
+	}
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			slog.Warn("Failed to close response body", "err", err)
+		}
+	}()
+	encoding := res.Header.Get("Content-Encoding")
+	originalHeader := res.Header
+	code := res.StatusCode
+
+	responses, token, errcode, err := processData(res.Body, encoding, responseFn)
+	if err != nil {
+		code = errcode
+	}
+	return encoding, originalHeader, responses, token, code, err
+}
+
+func getResponses(req *http.Request, cli *http.Client, responseFn processResponse) ([]json.RawMessage, http.Header, string, int, error) {
+	encoding, originalHeader, responses, token, originalCode, err := doRequest(req, cli, responseFn)
```

Comment:
nit: what do you think about having a struct with these parameters? that would be easier to update and extract data from. Something like `multipartResponse`